File: algorithms/golang/addTwoNumbers/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 *
 * https://leetcode.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (30.82%)
 * Total Accepted:    818.5K
 * Total Submissions: 2.7M
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * You are given two non-empty linked lists representing two non-negative
 * integers. The digits are stored in reverse order and each of their nodes
 * contain a single digit. Add the two numbers and return it as a linked list.
 * 
 * You may assume the two numbers do not contain any leading zero, except the
 * number 0 itself.
 * 
 * Example:
 * 
 * 
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 * Explanation: 342 + 465 = 807.
 * 
 * 
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	result := &ListNode{}
	carry := 0
	sum := 0
	curr := result
	for {
		if l1 == nil && l2 == nil  {
			if carry == 1 {
				curr.Next = &ListNode{}
				curr = curr.Next
				curr.Val = 1
			}
			break
		} 

		curr.Next = &ListNode{}
		curr = curr.Next

		v1 := 0
		v2 := 0
		if l1 != nil {
			v1 = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			v2 = l2.Val
			l2 = l2.Next
		}

		sum = v1 + v2 + carry
		curr.Val = sum % 10

		if sum >= 10 {
			carry = 1
		} else {
			carry = 0
		}
	}

	return result.Next
}

File: algorithms/java/src/linkedlist/2.AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry=0,sum=0,add=0;
        ListNode head=new ListNode();
        ListNode n=new ListNode(0,head);
        while(l1!=null || l2!=null){
            ListNode l3=new ListNode();
            if(l1!=null && l2!=null){
                add=l1.val+l2.val;
                l1=l1.next;
                l2=l2.next;
            }
            else if(l1!=null){
                add=l1.val;
                l1=l1.next;
            }
            else{
                add=l2.val;
                l2=l2.next;
            }

            sum=(add+carry)%10;
            carry=(add+carry)/10;

            if(head==null){
                head.val=sum;
                head.next=null;
            }
            else{
                // n=head;
                while(n.next!=null){
                    n=n.next;
                }
                l3.val=sum;
                l3.next=null;
                n.next=l3;

            }

        }
        if(carry==1){
            ListNode node=new ListNode();
            node.val=carry;
            node.next=null;
            while(n.next!=null)
            n=n.next;
            n.next=node;
        }

        return head.next;
    }
}
